presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.geospatial;
+
+import com.esri.core.geometry.Geometry;
+import com.esri.core.geometry.GeometryEngine;
+import com.esri.core.geometry.OperatorImportFromESRIShape;
+import com.esri.core.geometry.SpatialReference;
+import com.esri.core.geometry.ogc.OGCGeometry;
+import com.facebook.presto.spi.PrestoException;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.airlift.slice.Slices.EMPTY_SLICE;
+import static java.util.Objects.requireNonNull;
+
+public final class GeometryUtils
+{
+    public static final String POINT = "Point";
+    public static final String LINE_STRING = "LineString";
+    public static final String POLYGON = "Polygon";
+    public static final String MULTI_POINT = "MultiPoint";
+    public static final String MULTI_LINE_STRING = "MultiLineString";
+    public static final String MULTI_POLYGON = "MultiPolygon";
+    public static final int WKID_UNKNOWN = 0;
+
+    private static final int SIZE_WKID = 4;
+
+    private GeometryUtils() {}
+
+    public static OGCGeometry deserialize(Slice shape)
+    {
+        requireNonNull(shape, "shape is null");
+        int wkid = shape.getInt(0);
+        ByteBuffer shapeBuffer = getShapeBuffer(shape);
+        if (shapeBuffer.limit() < SIZE_WKID || shapeBuffer.getInt(0) == Geometry.Type.Unknown.value()) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Corrupted bytes for Geometry");
+        }
+
+        SpatialReference spatialReference = null;
+        if (wkid != WKID_UNKNOWN) {
+            spatialReference = SpatialReference.create(wkid);
+        }
+        Geometry esriGeom = OperatorImportFromESRIShape.local().execute(0, Geometry.Type.Unknown, shapeBuffer);
+        OGCGeometry createdGeom = OGCGeometry.createFromEsriGeometry(esriGeom, spatialReference);
+        return createdGeom;
+    }
+
+    public static Slice serialize(OGCGeometry input)
+    {
+        int wkid = input.SRID();
+        Geometry geometry = input.getEsriGeometry();
+
+        if (geometry == null) {
+            return EMPTY_SLICE;
+        }
+        byte[] shape = GeometryEngine.geometryToEsriShape(geometry);
+
+        if (shape == null) {
```

Comment:
Is this possible? Under what circumstances GeometryEngine.geometryToEsriShape(geometry) returns null for non-null input? Is this the case for empty geometries? Wouldn't it be clearer to check that before calling geometryToEsriShape?

Reply:
shape will never be null, will update